### PR TITLE
"Dual Arm Husky" addition

### PR DESCRIPTION
**Pull request by UltronDestroyer**
Adding it in raw and will make changes from there.

### CODE REVIEW

File: husky_bringup/udev/52-ftdi.rules [hunk 1]
```diff
@@ -1 +1,2 @@
 SUBSYSTEMS=="usb", ATTRS{idVendor}=="0403", ATTRS{idProduct}=="6001", MODE="0666", SYMLINK+="ftdi_%s{serial}"
+SUBSYSTEMS=="usb", ATTRS{idVendor}=="0403", ATTRS{idProduct}=="6015", MODE="0666", SYMLINK+="ftdi_%s{serial}"
```
Comment (UltronDestroyer) [inline husky_bringup/udev/52-ftdi.rules:2]: This allows us to add the robotiq force torques on udev

File: husky_bringup/scripts/install [hunk 2]
```diff
@@ -23,4 +23,22 @@ if os.environ.get('HUSKY_UR5_IP'):
...
+if os.environ.get('BUMBLEBEE_ENABLED'):
+  j.add(package="husky_bringup", glob="launch/bumblebee/*")
+
+if os.environ.get('ROBOTIQ_GRIPPERS_ENABLED'):
+  j.add(package="husky_bringup", glob="launch/robotiq_s_model_grippers/*")
+
+if os.environ.get('VELODYNE_VLP16_ENABLED'):
+  j.add(package="husky_bringup", glob="launch/velodyne/*")
+
+if os.environ.get("ROBOTIQ_FT_300_ENABLED"):
```
Comment (UltronDestroyer) [inline husky_bringup/scripts/install:41]: This adds a myriad of sensors available to be installed via the install script (for now ,it works only for the dual ur5 arm config, but in the future can be abstracted to set positions, xyz, rpy, for each respective sensor.. moving us to a more general robot configuration that is happily met by only touching env vars

File: husky_bringup/udev/52-ftdi.rules (hunk 1 above)
Comment (paulbovbel) [inline husky_bringup/udev/52-ftdi.rules:2]: use a regex for 6001/6015 matching in one stanza

File: husky_bringup/launch/robotiq_s_model_grippers/robotiq_s_gripper_left.launch [hunk 3]
```diff
@@ -0,0 +1,13 @@
+<launch>
```
Comment (paulbovbel) [inline husky_bringup/launch/robotiq_s_model_grippers/robotiq_s_gripper_left.launch:1]: parametrize and merge with gripper_right

File: husky_bringup/launch/ft_300/ft_300_right.launch [hunk 4]
```diff
@@ -0,0 +1,15 @@
+<launch>
+
```
Comment (paulbovbel) [inline husky_bringup/launch/ft_300/ft_300_right.launch:2]: indentation, extra spacing

File: husky_bringup/launch/ft_300/ft_300_right.launch [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+      <arg name="serial_id" default="ftdi_DA4N19U"/>
```
Comment (paulbovbel) [inline husky_bringup/launch/ft_300/ft_300_right.launch:4]: read port from optenv

File: husky_bringup/launch/ft_300/ft_300_right.launch [hunk 6]
```diff
@@ -0,0 +1,15 @@
+<launch>
```
Comment (paulbovbel) [inline husky_bringup/launch/ft_300/ft_300_right.launch:1]: parametrize, merge with ft_300_left

File: husky_bringup/launch/dual_ur5_arm_config/right_ur5.launch [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    <node name="right_ur5_driver" pkg="ur_modern_driver" type="ur_driver" args="$(arg robot_ip) $(arg reverse_port)" output="screen" respawn="true">
+      <param name="min_payload" type="double" value="$(arg min_payload)"/>
+      <param name="max_payload" type="double" value="$(arg max_payload)"/>
+      <param name="max_velocity" type="double" value="$(arg max_velocity)"/>
+      <param name="prefix" type="string" value="$(arg prefix)"/>
+      <param name="reverse_port" value="$(arg reverse_port)"/>
+    </node>
+
+    <!-- TF Buffer Server -->
+    <node pkg="tf2_ros" type="buffer_server" name="right_buffer_server" respawn="true">
```
Comment (paulbovbel) [inline husky_bringup/launch/dual_ur5_arm_config/right_ur5.launch:62]: why is this here

File: husky_bringup/launch/dual_ur5_arm_config/right_ur5.launch [hunk 8]
```diff
@@ -0,0 +1,69 @@
+<?xml version="1.0"?>
```
Comment (paulbovbel) [inline husky_bringup/launch/dual_ur5_arm_config/right_ur5.launch:1]: parametrize, merge wit left_ur5.launch and ur5.launch

File: husky_bringup/launch/dual_ur5_arm_config/husky_dual_ur5_planning_execution.launch [hunk 9]
```diff
@@ -0,0 +1,13 @@
+<launch>
```
Comment (paulbovbel) [inline husky_bringup/launch/dual_ur5_arm_config/husky_dual_ur5_planning_execution.launch:1]: this launch is unused.

Guessing it's robot/sim generic, so it should be in husky_control either way

Probably generalizable to single ur5 case as well

File: husky_bringup/launch/bumblebee/bumblebee.launch [hunk 10]
```diff
@@ -0,0 +1,38 @@
+<launch>
```
Comment (paulbovbel) [inline husky_bringup/launch/bumblebee/bumblebee.launch:1]: This seems redundant with https://github.com/ros-drivers/pointgrey_camera_driver/blob/master/pointgrey_camera_driver/launch/bumblebee.launch, should be arg'ed and include/reused as necessary

Comment (tonybaltovski) [inline husky_bringup/launch/bumblebee/bumblebee.launch:1]: I agree.  Should use environment variables since the serial which change per robot and other settings might need to be changed.  Check out the Jackal: https://github.com/jackal/jackal_robot/blob/indigo-devel/jackal_bringup/launch/accessories.launch#L18

File: husky_bringup/launch/flir_ptu/flir_ptu.launch [hunk 11]
```diff
@@ -0,0 +1,11 @@
+<launch>
+
+  <arg name="port" default="/dev/ttyS1"/>
```
Comment (tonybaltovski) [inline husky_bringup/launch/flir_ptu/flir_ptu.launch:3]: These will not be changeable, use environment variables.  There could be cases when just a flir_ptu is used.

File: husky_bringup/launch/ft_300/ft_300_left.launch [hunk 12]
```diff
@@ -0,0 +1,14 @@
+<launch>
+
+     
+      <arg name="serial_id" default="ftdi_DA4PE3R"/>
```
Comment (tonybaltovski) [inline husky_bringup/launch/ft_300/ft_300_left.launch:4]: environment variable

File: husky_bringup/launch/ft_300/ft_300_right.launch (hunk 5 above)
Comment (tonybaltovski) [inline husky_bringup/launch/ft_300/ft_300_right.launch:4]: environment variable

File: husky_bringup/launch/robotiq_s_model_grippers/robotiq_s_gripper_right.launch [hunk 13]
```diff
@@ -0,0 +1,11 @@
+<launch>
+
+<remap from="/husky_gripper_right/joint_states" to="/joint_states"/>
+
+  <group ns="husky_gripper_right"> 
+     <node pkg="robotiq_s_model_control" type="SModelTcpNode.py" respawn="true" args="192.168.1.19" output="screen" name="right_s_model_gripper_driver"/>
```
Comment (tonybaltovski) [inline husky_bringup/launch/robotiq_s_model_grippers/robotiq_s_gripper_right.launch:6]: environment variable for host IP.

File: husky_bringup/package.xml [hunk 14]
```diff
@@ -36,6 +36,7 @@
   <run_depend>tf2_ros</run_depend>
   <run_depend>um6</run_depend>
   <run_depend>ur_driver</run_depend>
+  <run_depend>ur_modern_driver</run_depend>
```
Comment (tonybaltovski) [inline husky_bringup/package.xml:39]: How about the dependencies for PointGrey cameras, flir_ptu, velodyne and robotiq?

File: husky_bringup/launch/robotiq_s_model_grippers/robotiq_s_gripper_left.launch (hunk 3 above)
Comment (UltronDestroyer) [inline husky_bringup/launch/robotiq_s_model_grippers/robotiq_s_gripper_left.launch:1]: Parametrizing WRT left and right won't work - these scripts are launched via /etc/ros/indigo/husky-core.d where parameters cannot be passed to the launch files without modification to the existing system. 

Henceforth, only a merge can be done by hard coding the two situations into one launch file.

File: husky_bringup/launch/ft_300/ft_300_right.launch (hunk 6 above)
Comment (UltronDestroyer) [inline husky_bringup/launch/ft_300/ft_300_right.launch:1]: For the same reason above - parametrization cannot be made here. merge is possible.

File: husky_bringup/launch/dual_ur5_arm_config/right_ur5.launch (hunk 7 above)
Comment (UltronDestroyer) [inline husky_bringup/launch/dual_ur5_arm_config/right_ur5.launch:62]: ? this gets installed using husky_bringup install and it is used to launch the UR5 arms for the husky. However we did determine that the user should launch these manually and so these should be moved somewhere else... But not sure where.... the existing scripts aren't namespaced, and don't use env vars.

File: husky_bringup/launch/dual_ur5_arm_config/husky_dual_ur5_planning_execution.launch (hunk 9 above)
Comment (UltronDestroyer) [inline husky_bringup/launch/dual_ur5_arm_config/husky_dual_ur5_planning_execution.launch:1]: This is here because when we boot the husky, I want it to start on boot. I'd rather have them separate, users who have a single ur5 enabled would get the husky_planning_execution from the install files there.

File: husky_bringup/launch/bumblebee/bumblebee.launch (hunk 10 above)
Comment (UltronDestroyer) [inline husky_bringup/launch/bumblebee/bumblebee.launch:1]: When including it in this location and running husky_bringup install when the env var is set, this will spawn on boot. 

I suppose however.. we could just tell the install script to look at the bumblebee.launch location. But then it will be pulling from source and the install script doesn't support env vars or args that we would need to plug into the launch file. That solution is out of scope of this PR tbh.

File: husky_bringup/launch/flir_ptu/flir_ptu.launch (hunk 11 above)
Comment (UltronDestroyer) [inline husky_bringup/launch/flir_ptu/flir_ptu.launch:3]: It's kind of a sticky situation. flir_ptu.launch is copied originally from the flir_ptu package, put in with our concrete parameters to the install directory. Users can pick what install launch flies to use with env vars.

However, we would need to make PRs to flir_ptu to correctly tell them to set env vars. Or modify the install scripts so that we can set env vars which get pushed down to launch file args.

File: husky_bringup/launch/robotiq_s_model_grippers/robotiq_s_gripper_left.launch (hunk 3 above)
Comment (paulbovbel) [inline husky_bringup/launch/robotiq_s_model_grippers/robotiq_s_gripper_left.launch:1]: Add a higher level launch file that does the include and parametrization, but DRY on the actual node launch and etc.

File: husky_bringup/launch/dual_ur5_arm_config/right_ur5.launch (hunk 7 above)
Comment (paulbovbel) [inline husky_bringup/launch/dual_ur5_arm_config/right_ur5.launch:62]: Why is the buffer server there. What uses it?

File: husky_bringup/launch/bumblebee/bumblebee.launch (hunk 10 above)
Comment (paulbovbel) [inline husky_bringup/launch/bumblebee/bumblebee.launch:1]: Chain to the full bumblebee launch from a small launch if necessary, but no need to reproduce ALL of it here.

**Comment by UltronDestroyer** [comment]
TODOs: (post-poning this merge until more time is allocated)
1. regex for udev rule for gripper and gps
2. Indentation, extra spacing
3. Optenv for robotiq gripper ports
4. merge gripper launch file, ft300 launch file
5. merge ur5 launch files.
6. env var for ft300 launch files (serial id name)
7. env var for robotiq gripper launch files (host ip)
8. dependencies for pointgrey cameras, flir_ptu, velodyne, robotiq

**Comment by tonybaltovski** [comment]
Any updates on this @TheDash?

**Comment by UltronDestroyer** [comment]
nada, I was moved away from this project after it shipped

**Comment by tonybaltovski** [comment]
@TheDash will close, reopen if needed.

**Comment by wxmerkt** [comment]
Any updates? Is there going to be a merge of this into the main branch? Who's on the project now?
